reflekt-plugin/src/test/kotlin/io/reflekt/plugin/ic/IncrementalCompilationTest.kt

```diff
@@ -0,0 +1,125 @@
+        Assertions.assertEquals(realExpectedResult, actualResult, "Result after IC is wrong!")
+
+        // Compare the initial result and result without IC
+        cacheDir.clear()
+        compileSources(cacheDir, srcRoots, compilerArgs, "Without IC")
+        val actualResultWithoutIC = runCompiledCode(outDir)
+        Assertions.assertEquals(realExpectedResult, actualResultWithoutIC, "The initial result and result after IC are different!")
+
+        testRoot.deleteRecursively()
+    }
+
+    private fun File.clear() {
+        this.deleteRecursively()
+        this.mkdir()
+    }
+
+    private fun runCompiledCode(outDir: File) = Util.runProcessBuilder(Util.Command(listOf("java", getMainClass(outDir)), directory = outDir.absolutePath))
+
+    // Find [mainFileName]Kt.class file in [outDir] and make the following transformations:
+    //  - сut <class> extension
+    //  - get the relative path with [outDir]
+    //  - replace all "/" into "."
+    private fun getMainClass(outDir: File): String {
+        val allFiles = FileUtil.getAllNestedFiles(outDir.absolutePath)
+        val mainClass = allFiles.find { it.name == "${mainFileName}Kt.class" }?.absolutePath?.removeSuffix(".class")
+            ?: error("The output directory doe not contains ${mainFileName}Kt.class file")
+        return mainClass.substring(mainClass.indexOf("$outFolderName/") + outFolderName.length + 1).replace("/", ".")
+    }
+
+    // If we had failed tests the previous results were not deleted and it can throw some compiler errors
+    private fun initTestRoot(): File {
+        val testRoot = File(getTempPath(), IncrementalCompilationTest::class.java.simpleName)
+        if (testRoot.exists()) {
+            testRoot.deleteRecursively()
+        }
+        testRoot.apply { mkdirs() }
+        return testRoot
+    }
+
+    private fun compileSources(cacheDir: File, sources: List<File>, compilerArgs: K2JVMCompilerArguments, errorMessagePrefix: String) {
+        val (_, errors) = compile(cacheDir, sources, compilerArgs)
+        check(errors.isEmpty()) { "$errorMessagePrefix build failed: \n${errors.joinToString("\n")}" }
+    }
+
+    private fun parseAdditionalArgs(testDir: File): List<String> {
+        return File(testDir, argumentsFileName)
+            .takeIf { it.exists() }
+            ?.readText()
+            ?.split(" ", "\n")
+            ?.filter { it.isNotBlank() }
+            ?: emptyList()
+    }
+
+    private fun applyModifications(sourcesDir: File, modifications: HashMap<File, Modification>): List<File> {
+        val allSources = FileUtil.getAllNestedFiles(sourcesDir.absolutePath)
+        val changedSources = mutableListOf<File>()
+        allSources.forEach { sourceFile ->
+            modifications.getOrDefault(sourceFile, null)?.let { m ->
+                m.applyActions()?.let {
+                    changedSources.add(it)
+                }
+            } ?: changedSources.add(sourceFile)
+        }
```

Comment:
since I think it is not convenient to have a hashmap with all files. I thought it is a god idea to store in the hashmap only files that have modifictaions